### PR TITLE
Swteam/#375/working back buttons

### LINKED ISSUE #375: Working "back"-buttons
**As a user** of the platform
**I want to** have intuitively working "back"-buttons
**In order to** navigate the platform easily.

## Acceptance Criteria:
 * It should be possible to get back from every subpage to its parent page without detours
    * example: event schedules and overview pages have no back button
 * Existing "back"-Buttons should refer to the expected page not the last visited page
    * example: Try editing an event and then clicking on back, currently you come back to the editing page but you should get back to the all events page
 * If you are not sure what the right page for referring to is, ask your PO :D

### Notes:
 * mentioned as bug #158 
 * It's probably wise to find a solution that works for all pages and does not require you to enter this data on every single page manually

### PR DESCRIPTION
**Pull request by Oluwoye**
There were major issues with Travis. Still needing feedback from PO.

Will close #375

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -170,7 +170,7 @@ end
 group :production do
   # Ruby interface to the PostgreSQL RDBMS
   # https://github.com/ged/ruby-pg
-  gem 'pg'
+  gem 'pg', '~> 0.18'
```
Comment (fpottbaecker) [inline Gemfile]: Why this change in version?

File: Gemfile.lock [hunk 2]
```diff
@@ -214,7 +214,6 @@ GEM
     parallel (1.12.1)
     parser (2.4.0.2)
       ast (~> 2.3)
-    pg (0.21.0)
```
Comment (fpottbaecker) [inline Gemfile.lock]: Why is this no longer present?

File: db/schema.rb [hunk 3]
```diff
@@ -82,13 +82,6 @@
     t.index ["user_id"], name: "index_organizers_on_user_id"
   end
 
-  create_table "participants", force: :cascade do |t|
```
Comment (fpottbaecker) [inline db/schema.rb]: Why is this table gone?

File: features/step_definitions/content.rb [hunk 4]
```diff
@@ -55,3 +55,33 @@
...
+    # There are a lot of routs in Rails, that can't be visited in general
+    begin
+      visit currentpath
+    rescue Exception
+      currentpath = '/imprint'
+      visit '/imprint'
+    end
+    puts currentpath
+    unless (excludedrouts.include? currentpath) || (temporarywedontknowwhattodorouts.include? currentpath)
+       expect(page).to have_link('Zurück')
```
Comment (fpottbaecker) [inline features/step_definitions/content.rb]: I18n.t please

File: features/back_buttons/present_buttons.feature [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  As a user of the platform
+  I want to have working "back" buttons
+  In order to navigate through the platform easily
+
+  Scenario: Visit all pages
+    Given a user
+    And the user is logged in
+    And a tournament 'MyTournamentus' with 4 teams
+    And a tournament 'Tourney2' with 2 users
+    And the Spielplan page for 'MyTournamentus' is visited
```
Comment (fpottbaecker) [inline features/back_buttons/present_buttons.feature]: This scenario looks a little bloated, why do we need to visit the tournaments page?

File: features/step_definitions/content.rb [hunk 6]
```diff
@@ -55,3 +55,33 @@
   }
   expect(page).to have_xpath("//h1[contains(text(),'#{I18n.t *translations[text]}')]")
 end
+
+
+Then(/^there should be a back button on all pages except the start page$/) do
```
Comment (fpottbaecker) [inline features/step_definitions/content.rb]: I think this should be an RSpec test so you can have a list of all relevant routes in your test ui.

**Review by fpottbaecker — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (Oluwoye) [reply Gemfile]: Because Travis load the 1.0.0 version automatically and failed with it. After I disabled pg at all, Travis said, that he wants to have version 0.18

File: db/schema.rb (hunk 3 above)
Comment (Oluwoye) [reply db/schema.rb]: Fixed

File: features/step_definitions/content.rb (hunk 6 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: We thought about this, but this is all functionality, that we might test: What do you think is best: Only a Cucumber test, only a RSpec test or a Cucumber and a redundant RSpec test?

File: features/back_buttons/present_buttons.feature (hunk 5 above)
Comment (Oluwoye) [reply features/back_buttons/present_buttons.feature]: To create at least one match and a complete schedule.

File: Gemfile.lock (hunk 2 above)
Comment (Oluwoye) [reply Gemfile.lock]: Fixed

File: features/step_definitions/content.rb (hunk 4 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Applied

File: app/views/users/edit_profile.html.erb [hunk 7]
```diff
@@ -0,0 +1,59 @@
+<%= render partial: "application/validation_errors", locals: {model: @user} %>
```
Comment (ErikBrendel) [inline app/views/users/edit_profile.html.erb]: why is there a whole new file?

File: app/views/events/ranking.html.erb [hunk 8]
```diff
@@ -34,3 +34,7 @@
     <% end %>
   </tbody>
 </table>
+
+<%= link_to t('helpers.links.back'),
```
Comment (ErikBrendel) [inline app/views/events/ranking.html.erb]: these calls are duplicating a bit. You could maybe extract this functionality to `ApplicationHelper::back_button_to(url)`

Would in this case look like `<%= back_button_to event_path(@event) %>`

File: features/step_definitions/content.rb [hunk 9]
```diff
@@ -55,3 +55,33 @@
...
+  Rails.application.routes.routes.each do |route|
+    currentpath = route.path.spec.to_s
+    if currentpath.include? '(.:format)'
+      currentpath['(.:format)'] = ''
+    end
+    if currentpath.include? ':id'
+      currentpath[':id'] = '1'
+    end
+
+    # There are a lot of routs in Rails, that can't be visited in general
```
Comment (ErikBrendel) [inline features/step_definitions/content.rb]: *`routes`

File: features/step_definitions/content.rb [hunk 10]
```diff
@@ -55,3 +55,33 @@
...
+    begin
+      visit currentpath
+    rescue Exception
+      currentpath = '/imprint'
```
Comment (ErikBrendel) [inline features/step_definitions/content.rb]: what?

File: features/step_definitions/content.rb [hunk 11]
```diff
@@ -55,3 +55,33 @@
...
+      visit '/imprint'
+    end
+    puts currentpath
```
Comment (ErikBrendel) [inline features/step_definitions/content.rb]: debug output should be removed

File: features/step_definitions/content.rb [hunk 12]
```diff
@@ -55,3 +55,33 @@
...
+    unless (excludedrouts.include? currentpath) || (temporarywedontknowwhattodorouts.include? currentpath)
+       expect(page).to have_link(I18n.t ('helpers.links.back'))
+    end
+  end;nil
```
Comment (ErikBrendel) [inline features/step_definitions/content.rb]: wow, what is that?

**Review by ErikBrendel — CHANGES_REQUESTED** [review]

File: app/views/users/edit_profile.html.erb (hunk 7 above)
Comment (Markus-Brand) [reply app/views/users/edit_profile.html.erb]: I think this file was deleted by us (merged with another file) and merge conflicts somehow recovered it.
You can just delete it, i guess

File: features/step_definitions/content.rb [hunk 13]
```diff
@@ -55,3 +55,33 @@
   }
   expect(page).to have_xpath("//h1[contains(text(),'#{I18n.t *translations[text]}')]")
 end
+
+
+Then(/^there should be a back button on all pages except the start page$/) do
+  # These routes were excluded in accordance with our PO
+  excludedrouts = ['/', '/cable', '/tournaments/new', '/leagues/new', '/leagues/1/edit', '/tournaments/1/edit', '/events/1/edit', '/rankinglists/new', '/teams/new', '/teams/1/edit', '/matches/1/edit', '/my/users/sign_in', '/my/users/password/edit', '/my/users/cancel', '/my/users/sign_up', '/users/1/dashboard', '/users/1/edit', '/matches/1/add_game_result', '/matches/1/edit_results']
```
Comment (Markus-Brand) [inline features/step_definitions/content.rb]: multiple cases of "routs" instead of "routes" here. Also please rename your variables to be seperated by underscores like this: excluded_routes

File: features/step_definitions/content.rb (hunk 10 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Can you explain your problem a bit more?

File: features/step_definitions/content.rb (hunk 12 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: The test searches all routes accessible in Rails. Not all of these routes are visitable with "visit page". But at all pages, which weren't excluded by PO and are visitable should be a back button.

File: features/step_definitions/content.rb (hunk 13 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Fixed

File: features/step_definitions/content.rb (hunk 9 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Yep done

File: features/step_definitions/content.rb (hunk 11 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Done

File: app/views/users/edit_profile.html.erb (hunk 7 above)
Comment (Oluwoye) [reply app/views/users/edit_profile.html.erb]: Me or Erik?^^ And should I remove this linke or the whole file? Where is  the new file for this page?

File: features/step_definitions/content.rb (hunk 10 above)
Comment (ErikBrendel) [reply features/step_definitions/content.rb]: why are you redirecting to the imprint in the error case? That does not make any sense to me

File: features/step_definitions/content.rb (hunk 12 above)
Comment (ErikBrendel) [reply features/step_definitions/content.rb]: I mean the `end;nil` part :D

File: features/step_definitions/content.rb (hunk 10 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Ohh good point! Fixed!

File: features/step_definitions/content.rb (hunk 12 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Ohh ... yep^^

Comment (Oluwoye) [reply features/step_definitions/content.rb]: Fixed ^^

File: features/step_definitions/content.rb (hunk 6 above)
Comment (fpottbaecker) [reply features/step_definitions/content.rb]: Ideally, just an rspec test, this seems to granular for an rspec test.

File: app/views/users/edit.html.erb [hunk 14]
```diff
@@ -120,6 +120,7 @@
       <%= link_to t('.cancel', default: t('helpers.links.cancel')),
                 user_path(@user), class: 'btn btn-default' %>
       <%= t('.unhappy') %>? <%= link_to t('.cancel_this_account'), user_path(@user), method: :delete %>.
+      <%= back_btn root_path %>
```
Comment (ErikBrendel) [inline app/views/users/edit.html.erb]: @Oluwoye I deleted `users/edit_profile`, since this file got merged into `users/edit` (as you see here). Is this one linge the only change you made to `users/edit_profile?`

**Review by ErikBrendel — APPROVED** [review]

File: app/views/users/edit.html.erb (hunk 14 above)
Comment (Oluwoye) [reply app/views/users/edit.html.erb]: yes

File: features/step_definitions/content.rb (hunk 6 above)
Comment (Oluwoye) [reply features/step_definitions/content.rb]: Do you mean "this seems too granular for a cucumber test"? Then I would copypaste all lines out of the step definitions, create a rspec with it and delete this test?

**Review by fpottbaecker — APPROVED** [review]

**Review by DerIntergalaktischeGrosshorst — APPROVED** [review]
